drivers/flash/flash_stm32_ospi.c

```diff
@@ -0,0 +1,1505 @@
+}
+
+/*
+ * Transfer Error callback.
+ */
+void HAL_OSPI_ErrorCallback(OSPI_HandleTypeDef *hospi)
+{
+	struct flash_stm32_ospi_data *dev_data =
+		CONTAINER_OF(hospi, struct flash_stm32_ospi_data, hospi);
+
+	LOG_DBG("Error cb");
+
+	dev_data->cmd_status = -EIO;
+
+	k_sem_give(&dev_data->sync);
+}
+
+/*
+ * Command completed callback.
+ */
+void HAL_OSPI_CmdCpltCallback(OSPI_HandleTypeDef *hospi)
+{
+	struct flash_stm32_ospi_data *dev_data =
+		CONTAINER_OF(hospi, struct flash_stm32_ospi_data, hospi);
+
+	LOG_DBG("Cmd Cplt cb");
+
+	k_sem_give(&dev_data->sync);
+}
+
+/*
+ * Rx Transfer completed callback.
+ */
+void HAL_OSPI_RxCpltCallback(OSPI_HandleTypeDef *hospi)
+{
+	struct flash_stm32_ospi_data *dev_data =
+		CONTAINER_OF(hospi, struct flash_stm32_ospi_data, hospi);
+
+	LOG_DBG("Rx Cplt cb");
+
+	k_sem_give(&dev_data->sync);
+}
+
+/*
+ * Tx Transfer completed callback.
+ */
+void HAL_OSPI_TxCpltCallback(OSPI_HandleTypeDef *hospi)
+{
+	struct flash_stm32_ospi_data *dev_data =
+		CONTAINER_OF(hospi, struct flash_stm32_ospi_data, hospi);
+
+	LOG_DBG("Tx Cplt cb");
+
+	k_sem_give(&dev_data->sync);
+}
+
+/*
+ * Status Match callback.
+ */
+void HAL_OSPI_StatusMatchCallback(OSPI_HandleTypeDef *hospi)
+{
+	struct flash_stm32_ospi_data *dev_data =
+		CONTAINER_OF(hospi, struct flash_stm32_ospi_data, hospi);
+
+	LOG_DBG("Status Match cb");
+
+	k_sem_give(&dev_data->sync);
+}
+
+/*
+ * Timeout callback.
+ */
+void HAL_OSPI_TimeOutCallback(OSPI_HandleTypeDef *hospi)
+{
+	struct flash_stm32_ospi_data *dev_data =
+		CONTAINER_OF(hospi, struct flash_stm32_ospi_data, hospi);
+
+	LOG_DBG("Timeout cb");
+
+	dev_data->cmd_status = -EIO;
+
+	k_sem_give(&dev_data->sync);
+}
+
+#if defined(CONFIG_FLASH_PAGE_LAYOUT)
+static void flash_stm32_ospi_pages_layout(const struct device *dev,
+				const struct flash_pages_layout **layout,
+				size_t *layout_size)
+{
+	struct flash_stm32_ospi_data *dev_data = DEV_DATA(dev);
+
+	*layout = &dev_data->layout;
+	*layout_size = 1;
+}
+#endif
+
+static const struct flash_driver_api flash_stm32_ospi_driver_api = {
+	.read = flash_stm32_ospi_read,
+	.write = flash_stm32_ospi_write,
+	.erase = flash_stm32_ospi_erase,
+	.get_parameters = flash_stm32_ospi_get_parameters,
+#if defined(CONFIG_FLASH_PAGE_LAYOUT)
+	.page_layout = flash_stm32_ospi_pages_layout,
+#endif
+};
+
+#if defined(CONFIG_FLASH_PAGE_LAYOUT)
+static int setup_pages_layout(const struct device *dev)
+{
+	const struct flash_stm32_ospi_config *dev_cfg = DEV_CFG(dev);
+	struct flash_stm32_ospi_data *data = DEV_DATA(dev);
+	const size_t flash_size = dev_cfg->flash_size;
+	uint32_t layout_page_size = data->page_size;
+	uint8_t value = 0;
+	int rv = 0;
+
+	/* Find the smallest erase size. */
+	for (size_t i = 0; i < ARRAY_SIZE(data->erase_types); ++i) {
+		const struct jesd216_erase_type *etp = &data->erase_types[i];
+
+		if ((etp->cmd != 0)
+		    && ((value == 0) || (etp->exp < value))) {
+			value = etp->exp;
+		}
+	}
+
+	uint32_t erase_size = BIT(value);
```

Comment:
Variables are generally declared at the start of the block.

Reply:
this function `setup_pages_layout()` is directly coming from the drivers/flash_controller/flash_stm32_qspi.c

Reply:
ok, ideally it should be fixed as well.

Reply:
Can we create a shared module for the common functions of QSPI and OSPI?